sockeye/coverage.py

```diff
@@ -129,6 +133,67 @@ def update_coverage(prev_hidden: mx.sym.Symbol,
         return update_coverage
 
 
+class FertilityCoverage(Coverage):
+    """
+    Coverage class that accumulates the attention weights for each source word,
+    and also computes a fertility value for each source word.
+    """
+
+    def __init__(self, max_fertility: int) -> None:
+        super().__init__()
+        self.max_fertility = max_fertility
+        # input (encoder) to fertility
+        self.cov_e2f_weight = mx.sym.Variable("%se2f_weight" % self.prefix)
+
+    def on(self, source: mx.sym.Symbol, source_length: mx.sym.Symbol, source_seq_len: int) -> Callable:
+        """
+        Returns callable to be used for updating coverage vectors in a sequence decoder.
+
+        :param source: Shape: (batch_size, seq_len, encoder_num_hidden).
+        :param source_length: Shape: (batch_size,).
+        :param source_seq_len: Maximum length of source sequences.
+        :return: Coverage callable.
+        """
+
+        # (batch_size, seq_len, 1)
+        source_fertility = mx.sym.FullyConnected(data=source,
+                                              weight=self.cov_e2f_weight,
+                                              no_bias=True,
+                                              num_hidden=1,
+                                              flatten=False,
+                                              name="%ssource_fertility_fc" % self.prefix)
+
+        def update_coverage(prev_hidden: mx.sym.Symbol,
+                            attention_prob_scores: mx.sym.Symbol,
+                            prev_coverage: mx.sym.Symbol):
+            """
+            :param prev_hidden: Previous hidden decoder state. Shape: (batch_size, decoder_num_hidden).
+            :param attention_prob_scores: Current attention scores. Shape: (batch_size, source_seq_len).
+            :param prev_coverage: Shape: (batch_size, source_seq_len, coverage_num_hidden).
+            :return: Updated coverage matrix . Shape: (batch_size, source_seq_len, coverage_num_hidden).
+            """
+
+            # (batch_size, seq_len, 1)
+            fertility = mx.sym.Activation(data=source_fertility,
+                                         act_type="sigmoid",
+                                         name="%sactivation" % self.prefix)
+
+            # (batch_size, seq_len, 1)
+            scaled_fertility = 1 / (self.max_fertility * fertility)
```

Comment:
could the activation and the scaling be moved outside `udpate_coverage()` and directly after the FullyConnected? They only have to be defined once per batch right?

Reply:
Good point - yes, if I understand correctly they can be computed once.